drivers/counter/counter_mcux_tpm.c

```diff
@@ -0,0 +1,303 @@
+#include <zephyr/logging/log.h>
+
+#include <fsl_tpm.h>
+
+LOG_MODULE_REGISTER(mcux_tpm, CONFIG_COUNTER_LOG_LEVEL);
+
+#define DEV_CFG(_dev) ((const struct mcux_tpm_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct mcux_tpm_data *)(_dev)->data)
+
+struct mcux_tpm_config {
+	struct counter_config_info info;
+
+	DEVICE_MMIO_NAMED_ROM(tpm_mmio);
+
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+
+	tpm_clock_source_t tpm_clock_source;
+	tpm_clock_prescale_t prescale;
+};
+
+struct mcux_tpm_data {
+	DEVICE_MMIO_NAMED_RAM(tpm_mmio);
+	counter_alarm_callback_t alarm_callback;
+	counter_top_callback_t top_callback;
+	uint32_t freq;
+	void *alarm_user_data;
+	void *top_user_data;
+};
+
+static TPM_Type *get_base(const struct device *dev)
+{
+	return (TPM_Type *)DEVICE_MMIO_NAMED_GET(dev, tpm_mmio);
+}
+
+static int mcux_tpm_start(const struct device *dev)
+{
+	const struct mcux_tpm_config *config = dev->config;
+	TPM_Type *base = get_base(dev);
+
+	TPM_StartTimer(base, config->tpm_clock_source);
+
+	return 0;
+}
+
+static int mcux_tpm_stop(const struct device *dev)
+{
+	TPM_Type *base = get_base(dev);
+
+	TPM_StopTimer(base);
+
+	return 0;
+}
+
+static int mcux_tpm_get_value(const struct device *dev, uint32_t *ticks)
+{
+	TPM_Type *base = get_base(dev);
+
+	*ticks = TPM_GetCurrentTimerCount(base);
+
+	return 0;
+}
+
+static int mcux_tpm_set_alarm(const struct device *dev, uint8_t chan_id,
+			      const struct counter_alarm_cfg *alarm_cfg)
+{
+	TPM_Type *base = get_base(dev);
+	uint32_t current = TPM_GetCurrentTimerCount(base);
+	uint32_t top_value = base->MOD;
+	struct mcux_tpm_data *data = dev->data;
+	uint32_t ticks = alarm_cfg->ticks;
+
+	if (chan_id != kTPM_Chnl_0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	if (ticks > (top_value)) {
+		return -EINVAL;
+	}
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
+		if (top_value - current >= ticks)
+			ticks += current;
+		else
+			ticks -= top_value - current;
+	}
+
+	if (data->alarm_callback) {
+		return -EBUSY;
+	}
+
+	data->alarm_callback = alarm_cfg->callback;
+	data->alarm_user_data = alarm_cfg->user_data;
+
+	TPM_SetupOutputCompare(base, kTPM_Chnl_0, kTPM_NoOutputSignal, ticks);
+	TPM_EnableInterrupts(base, kTPM_Chnl0InterruptEnable);
+
+	return 0;
+}
+
+static int mcux_tpm_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	TPM_Type *base = get_base(dev);
+	struct mcux_tpm_data *data = dev->data;
+
+	if (chan_id != kTPM_Chnl_0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	TPM_DisableInterrupts(base, kTPM_Chnl0InterruptEnable);
+	data->alarm_callback = NULL;
+
+	return 0;
+}
+
+void mcux_tpm_isr(const struct device *dev)
+{
+	TPM_Type *base = get_base(dev);
+	struct mcux_tpm_data *data = dev->data;
+	uint32_t current = TPM_GetCurrentTimerCount(base);
+	uint32_t status;
+
+	status =  TPM_GetStatusFlags(base) & (kTPM_Chnl0Flag | kTPM_TimeOverflowFlag);
+	TPM_ClearStatusFlags(base, status);
+	__DSB();
```

Comment:
I think we can use `z_barrier_dsync_fence_full` here to make this code architecture generic